### PR TITLE
Memory postgresml

### PR DESCRIPTION
**Pull request by santiatpml**
### Background

PostgresML can be used as an memory backend for Auto-GPT. PostgresML is an open-source machine learning extension to PostgreSQL that enables training and inference on text and tabular data. With PostgresML, you can integrate machine learning models into your PostgreSQL database. In addition, you can use your existing PostgreSQL database as a vector database by generating embeddings from text stored in the database.

Resources:
- [Generating LLM embeddings with open source models in PostgresML](https://postgresml.org/blog/generating-llm-embeddings-with-open-source-models-in-postgresml)
- [PostgresML as a Vector Database](https://github.com/postgresml/postgresml#vector-database)

### Changes

- `.env.template` includes a section on PostgresML environment variables
- `autogpt/config/config.py` includes PostgresML configuration
- `autogpt/memory/__init__.py` includes PostgresMLMemory as a supported memory backend
- `autogpt/memory/postgresml.py` A new Python file and a new class called `PostgresMLMemory` developed using `redis` implementation as a reference

### Documentation

- Each file that has been edited/added has respective inline comments
- Updated `docs/configuration/memory.md` with an additional section on PostgresML installation instructions and `pip` dependencies
- `pip` dependencies for PostgresML are not included in the requirements.txt

### Test Plan

These steps are included [in documentation](https://github.com/santiatpml/Auto-GPT/blob/memory-postgresml/docs/configuration/memory.md#postgresml-setup). Rewriting summary here:

1. Install PostgresML using Docker or spin up a PostgresML cloud database
2. Clone Auto-GPT and run `pip install -r requirements.txt`
3. To use PostgresML memory backend: 
    - Install libpq-dev : `sudo apt install libpq-dev`
    - `pip install psycopg2` 
4. `python -m autogpt -m postgresml` will use PostgresML as memory backend

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes 

I have not added tests related to this PR as this involves additional dependencies like spinning up a PostgresML database and Python packages. I have not seen any tests for redis or pinecone for reference. Happy to add any suggested tests.

### CODE REVIEW

**Comment by katmai** [comment]
any chance you can create a quality of life PR that would deal with taking redis data that would already be there, if we've run autogpt for a while now, and importing it seamlessly into PostgresML?

File: .env.template [hunk 1]
```diff
@@ -105,6 +105,13 @@ OPENAI_API_KEY=your-openai-api-key
 # MILVUS_SECURE=
 # MILVUS_COLLECTION=autogpt
 
+### PostgresML
```
Comment (amarkules1) [inline .env.template:109]: Could you add postgresML as a memory backend option under the section at line 51 above?

File: autogpt/memory/postgresml.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            data: The data to compare to.
+            num_relevant: The number of relevant data to return.
+
+        Returns: A list of the most relevant data.
+        """
+        query_embedding = create_embedding_with_ada(data)
+        query_embedding = ",".join(str(v) for v in query_embedding)
+        query_statement = (
+            "SELECT data FROM %s ORDER BY %s.embedding <-> ARRAY[%s]::vector LIMIT %d;"
+            % (self.postgresml_tablename, self.postgresml_tablename, query_embedding, 5)
```
Comment (amarkules1) [inline autogpt/memory/postgresml.py]: should the 5 here be replaced with `num_relevant`?

File: docs/configuration/memory.md [hunk 3]
```diff
@@ -139,8 +141,61 @@ WEAVIATE_API_KEY="your weaviate API key if you have one"
...
+
+Ubuntu: `sudo apt install libpq-dev`
+
+Centos/Fedora/Cygwin/Babun.: `sudo yum install libpq-devel`
+
+Mac: `brew install postgresql`
+
+**Install `psycopg2`**
+
+`pip install psycopg2`
```
Comment (amarkules1) [inline docs/configuration/memory.md:186]: pyscopg2 should be added to requirements.txt

File: .env.template (hunk 1 above)
Comment (santiatpml) [reply .env.template:109]: Done.

File: autogpt/memory/postgresml.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
+import psycopg2
+from psycopg2 import sql
+from typing import Any
+from autogpt.llm_utils import create_embedding_with_ada
```
Comment (amarkules1) [inline autogpt/memory/postgresml.py]: getting an error when trying to run your branch, doesn't look like this `create_embedding_with_ada` function exists.

Comment (amarkules1) [reply autogpt/memory/postgresml.py]: think you want `get_ada_embedding`

Comment (santiatpml) [reply autogpt/memory/postgresml.py]: I am updating my calls now. Will push a commit in a bit.

File: docs/configuration/memory.md (hunk 3 above)
Comment (santiatpml) [reply docs/configuration/memory.md:186]: `psycopg2` requires `libpq` to be installed. This might break `pip install -r requirements.txt`

**Comment by amarkules1** [comment]
just pulled your updates and everything seems to be working now, looks like you've got some merge conflicts to resolve though

**Review by amarkules1 — APPROVED** [review]

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Pull request closed without merging**